pkg/daemon/update.go

```diff
@@ -255,6 +207,76 @@ func (dn *Daemon) compareMachineConfig(oldConfig, newConfig *mcfgv1.MachineConfi
 	return true, nil
 }
 
+// addExtensionsRepo adds a repo into /etc/yum.repos.d/ which we use later to
+// install extensions and rt-kernel
+func addExtensionsRepo() error {
+	repoContent := "[coreos-extensions]\nenabled=1\nmetadata_expire=1m\nbaseurl=" + osImageContentDir + "/extensions/\ngpgcheck=0\nskip_if_unavailable=False\n"
+	if err := writeFileAtomicallyWithDefaults(extensionsRepo, []byte(repoContent)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func extractOSImage(imgURL string) error {
+	var registryConfig []string
+	if _, err := os.Stat(kubeletAuthFile); err == nil {
+		registryConfig = append(registryConfig, "--registry-config", kubeletAuthFile)
+	}
+	if err := os.MkdirAll(osImageContentDir, 0755); err != nil {
+		return fmt.Errorf("error creating directory %s: %v", osImageContentDir, err)
+	}
+	// Extract the image
+	args := []string{"image", "extract", "--path", "/:" + osImageContentDir}
+	args = append(args, registryConfig...)
+	args = append(args, imgURL)
+	if _, err := pivotutils.RunExt(numRetriesNetCommands, "oc", args...); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (dn *Daemon) applyOSChanges(oldConfig, newConfig *mcfgv1.MachineConfig) error {
+	// Extract image and add coreos-extensions repo if we have either OS update or package layering to perform
+	mcDiff, err := newMachineConfigDiff(oldConfig, newConfig)
+	if err != nil {
+		return err
+	}
+
+	if mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType {
+		if err := extractOSImage(newConfig.Spec.OSImageURL); err != nil {
+			return err
+		}
+		if dn.OperatingSystem == MachineConfigDaemonOSRHCOS {
+			if err := addExtensionsRepo(); err != nil {
```

Comment:
Why do we want to do this here rather than if/when we apply extensions in [`applyExtensions`](https://github.com/openshift/machine-config-operator/pull/1941/files#diff-06961b075f1753956d802ba954d2cfb5R747)

Reply:
Because extensions repo is also used to switch to or update kernel-rt packages. If you extract OSContainer, there is a extensions/ directory which has all additional rpm packages we ship and that includes kernel-rt packages as well. Maybe we can name the repo to something else but the actual repo directory which contains rpm has been named as extensions, so the name is in sync with that :)